Bio/PDB/__init__.py

```diff
@@ -88,3 +97,67 @@
     from .SASA import ShrakeRupley
 except ImportError:
     pass
+
+_FormatToParser = {"pdb": PDBParser(), "cif": MMCIFParser()}
+
+
+def read(handle, format, id=None):
+    """Parse a file or handle using one of the available parsers to generate a structure.
+
+    Arguments:
+     - handle    - handle to the file, or the filename as a string
+     - format    - string describing the file format. can be [pdb, cif, mmtf]
+     - id        - optional id of the structure, if None header information or then base name is used
+
+    If you have the file name in a string 'filename', use:
+
+    >>> from Bio import PDB
+    >>> filename = "1A8O.pdb"
+    >>> PDB.read(filename, format="pdb")
+
+    The function also supports file handles and gzip files
+
+    >>> from Bio import PDB
+    >>> PDB.read("1A8O.pdb.gz", format="pdb")
+    """
+    # TODO *kwargs for the parser to allow QUIET or permissive mode
+
+    # Try and give helpful error messages:
+    if not isinstance(format, str):
+        raise TypeError("Need a string for the file format (lower case)")
+    if not format:
+        raise ValueError("Format required (lower case string)")
+    if format != format.lower():
+        raise ValueError("Format string '%s' should be lower case" % format)
+
+    # need to use different parser for mmtf versus pdb,cif
+    if format == "mmtf":
+        # TODO: Raise warning if ID is set. MMTFParser does not support passing an id
+        # MMTF needs a seperate parser as get_structure does only take 1 positional argument and does not support file handles
+        if isinstance(handle, str):
+            parser = MMTFParser()
+            return parser.get_structure(handle)
+        else:
+            raise TypeError("'%s' parser only accepts str not handle" % format)
+
+    elif format in _FormatToParser:  # Map the file format to a parser
+        # if we are provided a path to a file, check if gzipped,
+        # else open handle directly
+        if isinstance(handle, str):
+            encoding = guess_type(handle)[1]  # uses file extension
+            _open = partial(gzip.open, mode="rt") if encoding == "gzip" else as_handle
+        else:
+            _open = as_handle
+
+        with _open(handle) as fp:
+            parser = _FormatToParser[format]
+            structure = parser.get_structure(id, fp)
+            # set id from header if given, else use basename
+            if structure.header["idcode"] != "" and id is None:
+                structure.id = structure.header["idcode"]
+            elif isinstance(handle, str) and id is None:
+                structure.id = Path(handle).stem
```

Comment:
Perhaps we can just leave `structure.id` as `None` in this case? Getting information from the file name seems a bit unusual.

Reply:
If a PDB file has no header it is most likely a PDB file generated by some kind of computational workflow so in that case I'd guess that the filename has some meaningful information.  So while unusual it might still be useful if loading multiple structures at the same time in order to identify the structures instead of having multiple structures with `None`.

Reply:
Wouldn't this lead to the situation that two files with identical contents but different filenames get different structures? For example, `1LCD.pdb` in the Biopython tests does not have an ID in its header. Then this could happen:
```
mdehoon ~/biopython-master/Tests/PDB 
$ cp 1LCD.pdb test.pdb

mdehoon ~/biopython-master/Tests/PDB 
$ python              
Python 3.9.1 (v3.9.1:1e5d33e9b9, Dec  7 2020, 12:44:01) 
[Clang 12.0.0 (clang-1200.0.32.27)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
```
```python
>>> from Bio import PDB
>>> s = PDB.read("1LCD.pdb", "pdb")
[... some unrelated warnings ...]
>>> t = PDB.read("test.pdb", "pdb")
[... some unrelated warnings ...]
>>> s == t
False
>>> s.id
'1LCD'
>>> t.id
'test'
>>> s.id = None
>>> t.id = None
>>> s == t
True
```

Reply:
Yes this can happen. The id is however only descriptive as far as I am aware so in real world applications I wouldn't say that it is undesired to indicate that the original file used to construct the structure was named differently.

Reply:
I'd echo @mdehoon's concern about treating the filename as meaningful. None seems a better choice.

Reply:
Actually I am wondering now what the purpose of the `id` attribute is.

If I parse a PDB file with a complete header line, like 1A8O.pdb in our tests, and I specify `"test"` as the `id`, then this is stored as `structure.id`:
```python
>>> from Bio.PDB import PDBParser
>>> parser = PDBParser()
>>> structure = parser.get_structure("test", "1A8O.pdb")
>>> structure.id
'test'
```
But the ID code from the header is also stored by the parser:
```python
>>> structure.header['idcode']
'1A8O'
```
Then what is the purpose of `id`? The `get_structure` docstring says
```
get_structure(id, file) method of Bio.PDB.PDBParser.PDBParser instance
    Return the structure.
    
    Arguments:
     - id - string, the id that will be used for the structure
     - file - name of the PDB file OR an open filehandle
```
but I could not understand the purpose of `id` from this. @JoaoRodrigues  do you know what the `id` attribute is good for?

Reply:
I never understood what the user-supplied ID was for when using the PDB module.

Reply:
As I commented above, I actually agree with @duerrsimon and would go with the filename as ID. The ID is only used when printing a `Structure` (to the command line). Other than that, it has no value. IDs of other Entities are meaningful though. That said, we could keep it as None but, I think the file name is a very good default value as well. Always helps to have a point to where the structure came from.

Reply:
If`Bio.PDB` does not need the `id` attribute, I think we should not set it at all (i.e. don't even set it to `None`). We don't store the filename in other Biopython parsers. As noted above, `__eq__` on identical structures stored in different files will return `False` ; this is very confusing. If it is the same file known under a different name, e.g. a soft-linked file, comparing the structures will also return `False`. Even if it is the same file under the same name, you may get a different `id` (for example, `structure = PDB.read('myfile.pdb')` and `handle = open('myfile.pdb'); structure = PDB.read(handle)` will give structures with different IDs. And if the `id` is the same, you still won't know if the structures came from the same file: They may come from different files that happen to have the same file name but are stored in different directories, or one stored locally and the other one remotely.

I can see that in specific cases it may be useful to store the filename as an attribute to the structure, depending on the circumstances and purpose in each case, but in general, I think this will open up endless opportunities for bugs. Users who do want to store the filename as an `id` attribute can always do so after `PDB.read` returns.

Note also that the `id` attribute on components of the structure (such as the individual chains) are read from the contents of the PDB file and are well-defined. There is some inconsistency in using the same attribute name for a well-defined property and for an essentially arbitrary property.

Reply:
Ok I get your concerns.  
So lets keep it at the following then? 
- if optional id is set use that as id
- if file contains id in header = use that as id  (will be consistent for identical files with different names)
- else set to None

Maybe it'd be worth to change __repr__ of structure so that id is only printed if set? 
printing  `<Structure>` instead of `<Structure id=None>` 
None of the unit tests fails if I implement this.

Reply:
Why complicate things, does the new ``Bio.PDB.read(...)`` function need an ``id=...`` argument at all? Surely just two mandatory arguments are enough for now: filename/handle and format.

Reply:
Comparing structures using `__eq__` is a bad idea. It's not even a well-defined question. For example, to begin with, do you want to compare coordinates or chemical composition?

I'd leave it as None to be honest (which is the equivalent to being not set).

Reply:
I agree with @JoaoRodrigues  that it is unclear what `__eq__` should be doing. But currently `__eq__` is  already defined for `Structure` objects in the `Entity` class from which `Structure` inherits. This `__eq__` implementation just checks whether the `id` attribute is the same. The same `__eq__` is used for other entities like chains and atoms. This definition of `__eq__` gives some surprising results, for example chains coming from different structures may be evaluated as equal:
```
>>> from Bio.PDB import PDBParser
>>> parser = PDBParser()
>>> structure1 = parser.get_structure('1A8O', "1A8O.pdb")
>>> structure2 = parser.get_structure('1LCD', "1LCD.pdb")
>>> chains1 = list(structure1.get_chains())
>>> chains2 = list(structure2.get_chains())
>>> chains1[0].id
'A'
>>> chains2[0].id
'B'
>>> chains2[1].id
'C'
>>> chains2[2].id
'A'
>>> chains1[0] == chains2[2]
True
```
Then I think it's fair to say that `__eq__` in `Bio.PDB` is broken.

I don't think `id` should be an argument to the `read` function, not even with default value `None`, because doing so will make it part of the official API, and it takes time to change that (having to declare it obsolete, and then two releases with a deprecation warning on it before we can change it).

To move forward, how about the following:
- define `read` as ` def read(handle, format):` without an `id` argument;
- if the file contains `id` in the header, use that as`id`;
- else set `id` to None;
- deprecate `__eq__` as well as accessing (but not setting) the `id` attribute.

Reply:
I agree with @mdehoon except on the very final point of deprecating accessing the ``id`` - it may be useful for those files which contain it.

Reply:
@peterjc  For those files, the ID is still accessible as `structure.header['idcode']`.
But I am OK to keep the `id` attribute as is for now, and discuss whether the `id` attribute should be deprecated in a separate PR.

Reply:
Appreciate the clarification. I'd remove any ID handling from the new ``read`` function (and leave the existing parsers to set it if they do currently).

Reply:
Just to clarify: `__eq__` is not broken. It's comparing labels of different components so it makes sense that if two different proteins chains have the same `id` they are equal. We never intended (afaik) to use `__eq__` for true equivalency between chemical structures, either coordinate or composition based. There is a lot of modeling in protein work so structures often don't have an "official" id, just the one you (user) assign them. Exceptions are, as you found, structures downloaded from the PDB, which do have an ID code. But even there there is redundancy (multiple PDB IDs point to the same protein and conformation). Also, I can tell you that in more than 10 years using Biopython, I _never_ used `structure.header["idcode"]` :)

The `Structure.id` is a by-product of `Entity` and inheritance. While you could argue that it's hard to define what it should be, compared to a `Chain.id` for example, and that it could be removed, it's still useful in practice for many cases. So, having a way to define when you load the structure would be handy.

Finally, I don't think a new function should introduce a new API that differs from the underlying code we already have. While we might want to move on to greener pastures, this is definitely a minor point that will only confuse our users. In the future, if people use `read` by default, we can rework the parsers behind the scenes and simply have deprecation warnings for these arguments as we remove them.

For these reasons, I'd leave the `id` in as an optional argument, with a default value of`None`.

Reply:
Since we haven't reached an agreement here, and we should not keep @duerrsimon waiting forever, let's defer the discussion of `__eq__` and the purpose and usage of the `id` attribute to a separate PR.

@peterjc 's suggestion:

> I'd remove any ID handling from the new read function (and leave the existing parsers to set it if they do currently).

then seems reasonable.

Reply:
OK with you @JoaoRodrigues?

It sounds like you have some real world use cases for setting the top level ID (something I have never needed to do while working with structures), which may argue for _later_ including this as an optional argument to the new ``Bio.PDB.read(...)`` function?

Reply:
I'd be fine with `Bio.PDB.read()` setting `id` to `None` internally but not exposing it as an optional argument for now!